secp256k1.nim

```diff
@@ -85,11 +94,11 @@ type
   SkMessage* = MDigest[SkMessageSize * 8]
     ## Message that can be signed or verified
 
-  SkEcdhSecret* = object
+  SkEcdhSecret* {.requiresInit.} = object
     ## Representation of ECDH shared secret
     data*: array[SkEdchSecretSize, byte]
 
-  SkEcdhRawSecret* = object
+  SkEcdhRawSecret* {.requiresInit.} = object
```

Comment:
You have to close the `default(T)` loophole by introducing the following overload:

```nim
func default(T: type SkEcdhRawSecret) {.error.}
```

Without it, the compiler will produce only a warning when `default(SkEcdhRawSecret)`` is encountered. Hopefully, one day Nim will become more strict about things and will start treating this as an error by default.

If this is deemed too much work, It's also possible to address it nim-stew by introducing a new type-section pragma that adds `requiresInit` and closes the `default` loophole in one go.

Reply:
hm, that might be appropriate - ie intro the macro and make sure all current code uses it so that whoever copy-pastes doesn't have to remember this loophole

Reply:
is there anything in 1.2 that can close the `clear` loophole as well?

Reply:
No, something like this may be possible with a [Z3 ghost variable](http://whiley.org/2014/06/20/understanding-ghost-variables-in-software-verification/) in a future release of Nim.

Reply:
in rust this would simply pass ownership of the instance to `clear` - I was hoping something like `sink` might work - z3 sounds like an awfully complicated solution

Reply:
Well, that's having support for [affine/linear types](https://en.wikipedia.org/wiki/Substructural_type_system) which is something I would also lobby for because I like their applications in smart contracts (e.g. the currency types in Libra's move language). These are not applicable in all situations though (the value needs to have an affine/linear type such as `unique_ptr` in the first place).

Ghost variables are arguably easier to explain than affine/linear types. You might be confusing familiarity with simplicity. They can also express many more things in more intuitive ways.

`sink` in particular is not about ownership. It's about the C++ practice of ["want speed? pass by value"](https://web.archive.org/web/20140205194657/http://cpp-next.com/archive/2009/08/want-speed-pass-by-value/).

Reply:
>

Reply:
The manual on destructors has a different opinion: https://github.com/nim-lang/Nim/wiki/Destructors#sink-parameters

> A sink parameter conveys a transfer of ownership. The parameter will be consumed.


> You might be confusing

you might be confusing the fact that natively supporting a feature in the language is more simple than referring the user to a theorem prover that is glued on to the language - wielding buzzwords like affine and linear types doesn't change this, because in order to understand what a ghost variable is, you must first learn how a theorem prover operates, what its limitations are etc - once you've come this far, you'll of course understand what ghost variables are, what make theorem provers fundamentally slow and limited and _why_ ghost variables solve the problem. 

Ownership from that point of view is easy: you own it or not, and if you don't, the compiler complains.

Reply:
The manual is using ambiguous language. `sink` is just an optimisation that will steal resources only if the the "sinked" variable has to further reads. Otherwise, it will perform a copy. In other words, exactly the C++ optimisation I've linked above.

Reply:
The ghost variables allow you to express things more intuitively, because they work pretty much like normal variables. You'll have a "field" specifying whether the value is in a cleared or valid state. 

There are many abstractions in computer science that you can use relatively easily without fully grasping how they work under the hood (e.g. take any crypto API as an example, or anything that has a "solver" in the name). The fact that there is a particular theorem prover involved and a particular algorithm used behind the scenes is not that much of an issue.

Reply:
Ambiguous?

> A sink parameter must be consumed exactly once within the proc's body.

"must be consumed exactly once` is in bold even. something is wrong here.

Reply:
This documentation needs updating. A `sink` parameter doesn't need to be consumed exactly once. It will be implicitly destroyed.

Reply:
It occurred to me that you are not citing the official documentation. The correct one is in here:

https://nim-lang.org/docs/destructors.html#sink-parameters

Reply:
so in theory a sink + a copy hook should limit things:

https://nim-lang.org/docs/destructors.html#lifetimeminustracking-hooks-eq-copy-hook